File: week_8_and_9/3_Ruby/rectangle_class/my_solution.rb
# U3.W8-9: Implement a Rectangle Class


# I worked on this challenge with Andy Principe.

# 2. Pseudocode
  # 1. Define area method which returns the area of the rectangle
  # 2. Define perimeter method which returns the perimeter of a rectangle (or square)
  # 3. Define diagonal method which returns the diagonal length of the rectangle (or square)
  # 4. Define method square?, which returns true if the rectangle is a square (if width = height)

# 3. Initial Solution
# class Rectangle
#   attr_accessor :width, :height

#   def initialize(width, height)
#     @width  = width
#     @height = height
#   end

#   def ==(other)
#     (other.width  == self.width && other.height == self.height ) ||
#     (other.height == self.width && other.width  == self.height )
#   end

#   def area
#     @width * @height
#   end

#   def perimeter
#     2*@width + 2*@height
#   end

#   def diagonal
#     Math.sqrt(@width**2 + @height**2)
#   end

#   def square?
#     @width == @height
#   end
# end




# 4. Refactored Solution

class Rectangle
  attr_accessor :width, :height

  def initialize(width, height)
    @width  = width
    @height = height
  end

  # def ==(other)
  #   (other.width  == self.width && other.height == self.height ) ||
  #   (other.height == self.width && other.width  == self.height )
  # end

  #Doesn't look like we needed the method above. Not sure why it was there to begin with?

  def area
    @width * @height
  end

  def perimeter
    2*@width + 2*@height
  end

  def diagonal
    Math.sqrt(@width**2 + @height**2)
  end

  def square?
    @width == @height
  end
end




# 1. DRIVER TESTS GO BELOW THIS LINE

def assert(message)
  raise message unless yield
end

rectangle = Rectangle.new(10,20)
square = Rectangle.new(20,20)

assert("Area method should be defined."){Rectangle.method_defined?(:area)}
assert("Area of rectangle should be 200"){rectangle.area == 200}
assert("Area of square should be 400"){square.area == 400}

assert("Perimeter method should be defined."){Rectangle.method_defined?(:perimeter)}
assert("Perimeter of rectangle should be 60"){rectangle.perimeter == 60}
assert("Perimeter of square should be 80"){square.perimeter == 80}

assert("Diagonal method should be defined."){Rectangle.method_defined?(:diagonal)}
assert("Diagonal of rectangle should be 22.360679774997898"){rectangle.diagonal == 22.360679774997898}
assert("Diagonal of square should be 28.284271247461902"){square.diagonal == 28.284271247461902}

assert("Square? method should be defined."){Rectangle.method_defined?(:square?)}
assert("Square? of rectangle should be false"){rectangle.square? == false}
assert("Square? of square should be true"){square.square?}



# 5. Reflection 
# This challenge proved to be fairly elementary. Andy and I solidified our knowledge of using assert statements, creating instance methods, and referring to instance variables within our methods. We also reserached the attr_accessor statement, and we were able to better understand its functionality and application.

File: week_8_and_9/2_JavaScript/translate_to_js/my_solution.js
// U3.W8-9: Challenge you're converting


// I worked on this challenge with Andy Principe

// Original Ruby code:
// class Rectangle
//   attr_accessor :width, :height

//   def initialize(width, height)
//     @width  = width
//     @height = height
//   end

//   def area
//   	@width * @height
//   end

//   def perimeter
//   	2*@width + 2*@height
//   end

//   def diagonal
//   	Math.sqrt(@width**2 + @height**2)
//   end

//   def square?
//   	@width == @height
//   end
// end

// 2. Pseudocode
// add width property
// add height property
// define area method, which returns area of rectangle
// define permiter method, which returns perimeter of rectangle/square
// define diagonal method, which returns diagonal length
// define square? method, which returns true if rectangle is a square

// 3. Initial Solution

// var rectangle = {
// 	width: 10,
// 	height: 20,
// 	area: function(){
// 		return rectangle.width * rectangle.height
// 	},
// 	perimeter: function(){
// 		return 2 * rectangle.width + 2 * rectangle.height
// 	},
// 	diagonal: function(){
// 		return Math.sqrt(Math.pow(rectangle.width,2) + Math.pow(rectangle.height,2))
// 	},
// 	is_square: function(){
// 		return rectangle.width === rectangle.height
// 	}
// };

// var square = {
// 	width: 20,
// 	height: 20,
// 	area: function(){
// 		return square.width * square.height
// 	},
// 	perimeter: function(){
// 		return 2 * square.width + 2 * square.height
// 	},
// 	diagonal: function(){
// 		return Math.sqrt(Math.pow(square.width,2) + Math.pow(square.height,2))
// 	},
// 	is_square: function(){
// 		return square.width === square.height
// 	}
// };

// 4. Refactored Solution


function Rectangle(width,height) {
	this.width = width;
	this.height = height;
	this.area = function(){
		return this.width * this.height
	};
	this.perimeter = function(){
		return 2 * this.width + 2 * this.height
	};
	this.diagonal = function(){
		return Math.sqrt(Math.pow(this.width,2) + Math.pow(this.height,2))
	};
	this.is_square = function(){
		return this.width === this.height
	}
};

var rectangle = new Rectangle(10,20);
var square = new Rectangle(20,20);


// 1. DRIVER TESTS/ASSERT STATEMENTS GO BELOW THIS LINE
function assert(test, message, test_number) {
  if (!test) {
    console.log(test_number + "false");
    throw "ERROR: " + message;
  }
  console.log(test_number + "true");
  return true;
}

assert(
  (rectangle instanceof Object),
  "The value of rectangle should be an Object.",
  "1. "
)

assert(
  (rectangle.width === 10),
  "The value of rectangle width should equal 10",
  "2. "
)

assert(
  (rectangle.height === 20),
  "The value of rectangle height should equal 20",
  "3. "
)

assert(
  (square instanceof Object),
  "The value of square should be an Object.",
  "4. "
)

assert(
  (square.width === 20),
  "The value of square width should equal 20",
  "5. "
)

assert(
  (square.height === 20),
  "The value of square height should equal 20",
  "6. "
)

assert(
  (rectangle.area() === 200),
  "The value of rectangle area should equal 200",
  "7. "
)

assert(
  (rectangle.perimeter() === 60),
  "The value of rectangle perimeter should equal 60",
  "8. "
)

assert(
  (rectangle.diagonal() === 22.360679774997898),
  "The value of rectangle diagonal should equal 22.360679774997898",
  "9. "
)

assert(
  (rectangle.is_square() === false),
  "The value of rectangle square? should be false",
  "10. "
)

assert(
  (square.area() === 400),
  "The value of square area should equal 400",
  "11. "
)

assert(
  (square.perimeter() === 80),
  "The value of square perimeter should equal 80",
  "12. "
)

assert(
  (square.diagonal() === 28.284271247461902),
  "The value of square diagonal should equal 28.284271247461902",
  "13. "
)

assert(
  (square.is_square() === true),
  "The value of square square? should be false",
  "14. "
)


// 5. Reflection 
// This challenge was helpful in seeing the similarites and differences between JavaScript and Ruby. Although there were different syntactic obstacles to overcome, we were able to efficiently recreate our code as JS. In the initial solution, we created 2 separate objects, with basically the same code. We Refactored to make our code more DRY by creating a function and passing in width and height to more easily create the separate objects. 